### DISCUSSION TITLE
How to run system specs without recompiling assets every-time

### DISCUSSION THREAD

**Discussion by templeman15** [ElMassimo/vite_ruby#136] [Q&A] [answered]
@ElMassimo first off thank you for all your work, vite_ruby is amazing!!

I am reaching in regards to this:

First on the list is an integration test that can trigger a Hot Reload, and check the updated style or code.

I'm trying to run my system specs (Capybara / RSpec) and every time the assets need to be compiled and fail the system spec with a NetTimeout. After the assets compile I can run the tests successfully. This flow is not opportune at all however, I'd really love to see hot reloading take place (maybe running bin/vite --mode=test or something along those lines so the assets don't need to go through the full recompilation process before being able to run the specs.

Do you have any idea how to accomplish this? I don't see much at all in Vite or Vite Ruby (or google) that go into any depth on testing.

Thanks!

**Comment by ElMassimo** [comment]
__Edit__: now possible, see [__this answer__](https://github.com/ElMassimo/vite_ruby/discussions/136#discussioncomment-1559382).

---

I haven't tried it, but Vite.js added a `watch` mode that might do the trick.

Run `bin/vite build -- --watch` to start [the Rollup watcher](https://vitejs.dev/guide/build.html#rebuild-on-files-changes).

If using [`vite_ruby@3.0.1`](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/CHANGELOG.md#301-2021-10-27) or higher, you can run `bin/vite build --watch`.

Then, make sure that [`autoBuild`](https://vite-ruby.netlify.app/config/#autobuild) is disabled or it will attempt to rebuild assets, since the build process never finished and it never recorded a success/failure.

---

It should be technically possible to run tests with a Vite.js server running, but it's not advisable because you wouldn't test the behavior of the compiled JS in production.

```bash
bin/vite dev
# and then override the Rails environment:
VITE_RUBY_MODE=development bin/rspec
```

**Reply by templeman15** [reply]
@ElMassimo following up here with your recommendations. 

1. `bin/vite build -- --watch`

This does speed up the time it takes to compile the app but still takes 30-50 seconds. For some reason the changes also don't seem to be reflected in the specs, still shows outdated code in the html (taken by a screen-capture of the failing test).

2.

```
bin/vite dev
# and then override the Rails environment:
VITE_RUBY_MODE=development bin/rspec`
```

I understand what you're saying about not being comparable to production compiled assets. It looks like webpack might be doing something similar but did not take this long with webpack. This is really being a hinderance to being able to do TDD or really any style of testing during development.

Current process:

1) Write test
2) Test fails
3) Write new code 
4) Run spec
5) Vite starts compiling 
6) Spec fails because Vite is taking too long to compile (60-90) seconds
7) Run spec again after Vite finishes
8) Spec passes

It really seems like we need to be able to do it in these steps:

1) Write test
2) Test fails
3) Write new code 
4) Run spec
5) Spec passes

Regarding running against dev with `VITE_RUBY_MODE=development bin/rspec`

I get the following:

<img width="1067" alt="Screen Shot 2021-10-27 at 4 28 12 PM" src="https://user-images.githubusercontent.com/3027970/139161894-9e460b2e-05af-4234-8817-d54ba4c5128a.png">

Thank you for your quick response! I'm curious how no one else in the Vite community has not ran into a similar problem?

**Reply by ElMassimo** [reply]
Webpack needs to recompile changes as well.

It's unusual for a Vite.js build to take 50 seconds. A couple of questions:

- Are you explicitly setting `build.minify` to `terser`?

  The default is `esbuild`, and it's faster.

- Are you using `@vitejs/plugin-legacy`?
 
  You could try skipping the plugin in the test environment, it should be significantly faster without it.

- How much memory is the node process getting?

  Try bumping it to see if the build time decreases: `export NODE_OPTIONS=--max_old_space_size=4096`

---

> For some reason the changes also don't seem to be reflected in the specs

That [makes sense](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/manifest.rb#L115). If the performance of `--watch` were acceptable, this is something that could be solved with a separate setting, or by always refreshing the manifest in the `development` and `test` environments.

--- 

> Regarding running against dev with VITE_RUBY_MODE=development bin/rspec, I get a socket error

It would be interesting to figure out if there's anything in the `test` configuration of Rails that would cause this to happen. Based on the traces, [this path is being executed](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/dev_server_proxy.rb#L20), so the dev server was up and running and the request would be forwarded.

**Reply by templeman15** [reply]
Yeah it does seem like webpack does that, however, I can make code changes and re-run my specs and webpack doesn't take to long cause my specs pass pretty quickly. 

> Are you explicitly setting build.minify to terser?

I am not, I'm just using the default

> Are you using @vitejs/plugin-legacy?

I am not

> How much memory is the node process getting?

This seemed to have slowed it down, so I'm not sure how much it was getting beforehand. But it's taking 2-2.5 minutes now.

>  If the performance of --watch were acceptable

Unfortunately I don't think it will be acceptable. I think this is the only deal breaker from moving over to Vite. I just completed nearly everything else and this is the only thing I can't figure out.

> It would be interesting to figure out if there's anything in the test configuration of Rails that would cause this to happen.

It's very possible, I'll have to look into this approach more, but like you said I don't know if using the development assets would be wise. 


------------


Another idea I had was:

`bin/vite dev --mode=test `

With vite.config like:

```
  "test": {
    "autoBuild": false,
    "publicOutputDir": "vite-test",
    "port": 3037
  }
```

and just running the spec like:

` bundle exec rspec spec/system/some_spec_file.rb`

But with this changes are not picked up by the vite dev server (no HMR or any output by Vite with code changes) and RSPec is not looking at the dev server that is running, it's just trying to look in `public/vite-test`.

**Reply by templeman15** [reply]
I have also tried out [this](https://github.com/mprt-org/rollup-plugin-incremental) plugin.. It seemed promising but failed to render the chunks if we had this setting `preserveModules: true,` which it says is required.

**Reply by ElMassimo** [reply]
How does it fail? That seems promising if the build is taking too long.

Thanks for replying above. Are you using React?

Using Vue the transforms are really fast because they avoid a full babel AST parse, so I haven't run across a project that takes that long to build.

**Reply by templeman15** [reply]
Thanks for responding back!

It just gets to the point where it usually says rendering chunks in the build process but hangs there. With --verbose it also gives no additional output. 

We have a very large monolith that used webpack. It was taking 30-60 seconds to do incremental builds for every code change. Vite gets that to 50-100ms, which is fantastic! But this is the only thing hinging us on releasing it to the team. We can't really write tests effectively and run the tests as we are developing. 

In the output of Vite we are getting css (scss), JavaScript (react), JavaScript (svelte), fonts, SVGs, vendor (which is very large).

**Reply by ElMassimo** [reply]
Gotcha, have you tried profiling which plugins are taking longer?

I've thought about it, and I think it would be nice to leverage the Vite dev server when running tests locally—it's a reasonable trade-off as long as the CI does a production build, and it could fallback to autoBuild if the server is not available.

If you prepare a repo with some of the libraries and Vite plugins you are using (such as React and Svelte), I can take a look and see what's necessary in terms of environment flags to get HMR working on Rails with RSpec, and supporting it as a mode in Vite Ruby.

**Reply by templeman15** [reply]
I really appreciate it! If you need any assistance and can point me in the right direction I can open a PR if needed! I do think however making testing front of mind will be big for the community, particularly in regard to Ruby / Rails and Vite!

I will try and spin up a repo at least with rails and react how we have it setup tomorrow or over the weekend! I also agree with the trade offs you mentioned, ideally just running one Vite dev server and not one for dev and test. 

I will also try profiling the plugins. Is there a way of doing that during the build process or what do you recommend?

**Comment by ElMassimo** [comment, marked as answer]
Using a Vite dev server when running tests is now [supported out of the box](https://vite-ruby.netlify.app/guide/development.html#hmr-for-integration-tests-%E2%9C%85), starting from `vite_ruby@3.0.2`.

Added [a test in the example Rails app](https://github.com/ElMassimo/vite_ruby/blob/main/examples/rails/spec/features/home_spec.rb#L15-L16) that rewrites files and verifies that HMR is working.

@templeman15 Run `bin/vite upgrade`, give it a try, and let me know how it goes 😃 

---

__Related__: Found that [sourcemaps were generated when building in test mode](https://github.com/ElMassimo/vite_ruby/commit/2b4faac7b811af089f981861bcf7e3a3fd8486b2), which is unnecessary and would really slow down the build in an app like the one you described. Running `bin/vite upgrade` should also upgrade to `vite-plugin-ruby@3.0.3` which includes this fix.

**Reply by templeman15** [reply]
@ElMassimo this is amazing!!! Thank you so much for making the changes that quickly! 

I have two questions / pieces of feedback:

1) When first starting the dev server with the updated configuration I'm still gettin a timeout when running the specs. I can only assuming that it's doing it's first run of precompiling all of the assets (there is no output indicating such however, it just shows the dev server running). After a little bit of a time I can start using the HMR updates with specs and it works perfectly!

2) I reviewed your code changes and only one thing stood out to me and that is [here](https://github.com/ElMassimo/vite_ruby/commit/e253bba26d164aabc7a9526df504c207ad2cf6f9#diff-0b8c13d0a63a7815c29241df645e39e6db8ba6424621121811f5d6a121f0380eR100).. Would this mean that by default all CI environments will run in this new HMR development mode? I'm only assuming that because most people will not have a `CI` variable set in their CI by default. It might be wise to flip it where locally you add an environment variable if you want the behavior and on CI it's opted out by default?

**Reply by ElMassimo** [reply]
1. Something might be off with the configuration. Try the [example app](https://github.com/ElMassimo/vite_ruby/tree/main/examples/rails), you can run `bin/rspec` and it won't trigger a build as long as you have [previously](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/manifest.rb#L95-L99) started the Vite server with `bin/vite dev --mode=test`.

2. Even without the `CI` flag (which vendors such as CircleCI, Travis, Codeship, etc, all set by default), if no Vite server is running, it will fallback to `autoBuild`, and if assets had been precompiled in a previous step, it will skip the build and just read the manifest.

**Reply by templeman15** [reply]
1. I will try that!

2. Ahh that's great! To clarify, I'lll need `autoBuild: true` in test for that behavior?

**Reply by ElMassimo** [reply]
@templeman15 Yeah, which [comes by default](https://vite-ruby.netlify.app/config/#shared-configuration-file-%F0%9F%93%84) in the installation process.

Regarding the lack of output when the build is triggered during tests, have in mind that [Vite Ruby uses the Rails logger](https://github.com/ElMassimo/vite_ruby/blob/main/vite_rails/lib/vite_rails/engine.rb#L20-L24), which in the test environment it's often configured to output to a file instead of the terminal (`log/test.log`).

If you want to see the output, you could [manually assign a stdout logger](https://github.com/ElMassimo/vite_ruby/blob/main/examples/rails/spec/spec_helper.rb#L11) to `ViteRuby.instance.logger` in `spec_helper.rb`, or change the default behavior of the Rails logger in the test environment.

**Reply by templeman15** [reply]
Awesome, thank you for all your help!

**Reply by templeman15** [reply]
Just updating you after getting everything wrapped up.

I did not end up running just one Vite dev server. This took env vars for development and placed them into the test environment. I instead created an npm script along with the `concurrently` package to run both dev servers in one terminal session. 

`yarn add concurrently`

package.json:

`"vite": "concurrently -n vite-dev,vite-test \"bin/vite dev\" \"bin/vite dev --mode=test\""`

-------

I'm getting a lot of this output now (for source maps that were turned off): 

<img width="1386" alt="Screen Shot 2021-10-29 at 4 07 42 PM" src="https://user-images.githubusercontent.com/3027970/139517717-d160429b-4be6-4a57-a5d0-5fa8c773fc12.png">

----- 

I ran into a lot of problems pushing to CI with using `react_on_rails` because it's set up for webpack. I finally solved it by updating `react_on_rails` and setting `REACT_ON_RAILS_PRECOMPILE=false` on CI. 

Thanks again!

**Reply by ElMassimo** [reply]
Thanks for sharing!

In case it's useful, my favorite way to run several processes is [overmind](https://github.com/DarthSim/overmind), which is Procfile-based, and automatically creates a tmux session with one process per tab that you can connect to independently to see output or interactively debug.

[Example](https://github.com/ElMassimo/vite_ruby/blob/main/examples/rails/Procfile.dev), which I run using a `dev` global alias (`overmind s -f Procfile.dev`) that works on every repo with a dev Procfile.

You could extend the idea to a Procfile.test, to start the necessary Vite test server.

**Comment by zealot128** [comment]
I "Fixed" the Timeout error, by simply running the build once before all system specs:

```ruby
RSpec.configure do
  config.before(:all, type: :system) do
    puts "\e[31mPrecompiling assets for system specs \e[0m"
    ViteRuby.instance.manifest.resolve_entries('application', type: 'js')
    puts "\e[32m-> Finished! precompiling assets for system specs \e[0m"
  end
end
```

That fixes the Timeout Issues for me (Using Cuprite the Ferrum::TimeoutError, Ferrum::PendingConnectionsError errors).

**Reply by ElMassimo** [reply]
You should prefer `ViteRuby.commands.build` instead, as it's more explicit. The build will be skipped if it's up-to-date.

Additionally, when running integration tests in the CI, it might be preferable to run `bin/rake assets:precompile` before running specs (separates time spent precompiling assets from time spent running tests).

**Reply by zealot128** [reply]
> You should prefer `ViteRuby.commands.build` instead, as it's more explicit. The build will be skipped if it's up-to-date.
> 
> Additionally, when running integration tests in the CI, it might be preferable to run `bin/rake assets:precompile` before running specs (separates time spent precompiling assets from time spent running tests).

@ElMassimo Thanks for the info. Did not find the command and just read the Tag Helper where I found the resolve_entries command. I will use the one you wrote.
Yes, in CI we also run a ``rails assets:precompile`` beforehand. But locally it is often frustrating when you run a single spec and you get a Timeout once everytime you modify the assets. Before, I event had a rescue clause, that catches the Timeout, and rerun the ``visit`` call in the first spec, which was tedious.

**Reply by ElMassimo** [reply]
The way I deal with it locally is by using a different Cuprite timeout in development:

```ruby
  # Unlike the capybara timeout, which should be very short to ensure test fail
  # fast, this timeout needs to be long to ensure we give enough time for assets
  # to be precompiled.
  #
  # You can increase this value in `.env.development.local` in slower matchines.
  cuprite_timeout = ENV.fetch("CUPRITE_TIMEOUT", 30).to_i
```

and then:

```ruby
    timeout: cuprite_timeout,
    process_timeout: cuprite_timeout,
```

while keeping the Capybara timeout strict:

```rb
Capybara.configure do |config|
  config.default_max_wait_time = 3
end
```

But being explicit and running a build for system specs makes sense. It might be a good idea to create a `Testing` section in the Vite Ruby guide.
